src/boilerplate.rs

```diff
@@ -0,0 +1,236 @@
+//! Replaces <!--BOILERPLATE filename.inc--> comments.
+//! These can either be comment nodes (in which case the resulting fragment will
+//! be inserted), or the complete value of an element's attribute (in which case
+//! the text will become the attribute value).
+
+use std::io;
+use std::path::{Path, PathBuf};
+
+use html5ever::tendril::{self, SendTendril, StrTendril};
+use html5ever::{local_name, Attribute, LocalName, QualName};
+use markup5ever_rcdom::{Handle, NodeData};
+use tokio::fs::File;
+use tokio::task::JoinHandle;
+
+use crate::dom_utils::NodeHandleExt;
+use crate::parser;
+
+type SendStrTendril = SendTendril<tendril::fmt::UTF8>;
+
+enum Edit {
+    ReplaceHTML(Handle, JoinHandle<io::Result<File>>),
+    ReplaceAttr(Handle, QualName, JoinHandle<io::Result<SendStrTendril>>),
+    ReplaceText(Handle, JoinHandle<io::Result<SendStrTendril>>),
+}
+
+pub struct Processor {
+    /// Path to look for boilerplate files.
+    path: PathBuf,
+
+    /// Path to look for example files.
+    example_path: PathBuf,
+
+    /// Changes to be made in the apply step.
+    edits: Vec<Edit>,
+}
+
+impl Processor {
+    pub fn new(path: impl Into<PathBuf>, example_path: impl Into<PathBuf>) -> Self {
+        Self {
+            path: path.into(),
+            example_path: example_path.into(),
+            edits: vec![],
+        }
+    }
+
+    /// Should be called for each node in the document.
+    /// Identifies replacements which will be needed, and starts the necessary
+    /// I/O.
+    pub fn visit(&mut self, node: &Handle) {
+        match node.data {
+            // BOILERPLATE comments will need to be replaced with their
+            // corresponding HTML, parsed. Open the file so that we can do so on
+            // demand.
+            NodeData::Comment { ref contents } if contents.starts_with("BOILERPLATE ") => {
+                let path = Path::new(contents[12..].trim());
+                if is_safe_path(path) {
```

Comment:
Why do we have this is_safe_path check? If we omitted it, what bad thing would happen? Could we assert it, instead of silently skipping the comment if it fails? (Here and below.)